### PR TITLE
Return a generic array in Query

### LINKED ISSUE #377: QueryBuilder->getQuery() return type conflicts with Psalm
Since a recent release of the Psalm-doctrine plugin (https://github.com/psalm/psalm-plugin-doctrine/pull/115) the following code conflicts with Phpstan.

Can the new TKey parameter also be supported by the phpstan-docrine plugin? It looks like Psalm is more correct here.

Code:
```php
/** @var Query<array-key, Country> $query */
$query = $this->createQueryBuilder('country')->getQuery();
```

Without TKey (`/** @var Query<Country> $query */)` produces the following error in Psalm:
```
psalm: InvalidTemplateParam: Extended template param TKey of Doctrine\ORM\Query<App\Entity\Country\Country> expects type array-key, type App\Entity\Country\Country given
psalm: MissingTemplateParam: Doctrine\ORM\Query has missing template params, expecting 2
```

With TKey (`/** @var Query<array-key, Country> $query */`) procesed the following error in PHPStan:
```
phpstan: Generic type Doctrine\ORM\Query<(int|string), App\Entity\Country\Country> in PHPDoc tag @var for variable $query specifies 2 template types, but class Doctrine\ORM\Query supports only 1: TResult
```

### PR DESCRIPTION
**Pull request by annuh**
Fixes #377.

Same as https://github.com/psalm/psalm-plugin-doctrine/pull/115, but for PHPStan 🙂

### CODE REVIEW

File: stubs/ORM/Query.stub [hunk 1]
```diff
@@ -3,10 +3,24 @@
...
 
 /**
+ * @template TKey as array-key
  * @template TResult The type of results returned by this query in HYDRATE_OBJECT mode
  *
  * @extends AbstractQuery<TResult>
  */
 final class Query extends AbstractQuery
 {
+    public const HYDRATE_OBJECT = 1;
```
Comment (annuh) [inline stubs/ORM/Query.stub:13]: This was needed to fix:
```
 ------ ------------------------------------------------ 
  Line   stubs/ORM/Query.stub                            
 ------ ------------------------------------------------ 
  23     PHPDoc tag @return contains unresolvable type.  
 ------ ------------------------------------------------ 
```

File: tests/Type/Doctrine/data/QueryResult/queryResult.php [hunk 2]
```diff
@@ -232,36 +232,36 @@ public function testReturnTypeOfQueryMethodsWithExplicitNonConstantHydrationMode
...
 	 * VoidType
 	 *
-	 * @param Query<mixed> $query
+	 * @param Query<array-key, mixed> $query
 	 */
 	public function testReturnTypeOfQueryMethodsWithReturnTypeIsMixed(EntityManagerInterface $em, Query $query): void
 	{
 		assertType(
-			'mixed',
+			'array<(int|string)>',
```
Comment (annuh) [inline tests/Type/Doctrine/data/QueryResult/queryResult.php:240]: TBH, I'm not really sure about these changes but all the tests pass now 🙄

File: Makefile [hunk 3]
```diff
@@ -23,6 +23,6 @@ cs-fix:
 phpstan:
 	php vendor/bin/phpstan analyse -l 8 -c phpstan.neon src tests
 
-.PHONY: phpstan-generate-baseline
+.PHONY: phpstan-generate-baselmake ine
```
Comment (janedbal) [inline Makefile:26]: typo

File: stubs/ORM/Query.stub [hunk 4]
```diff
@@ -3,10 +3,24 @@
 namespace Doctrine\ORM;
 
 /**
+ * @template TKey as array-key
```
Comment (janedbal) [inline stubs/ORM/Query.stub:6]: This might be conflicting with the solution propsed in https://github.com/phpstan/phpstan-doctrine/pull/384

**Comment by ondrejmirtes** [comment]
Superseded by https://github.com/phpstan/phpstan-doctrine/commit/4490e56f6fe3f8985a03be073d4545c3fb59cdbb

**Pull request closed without merging**